recent SIMD changes broke Lambda

```

/home/mi/h4nn3s/devel/lambda/include/seqan/include/seqan/modifier/modifier_padding.h: In instantiation of 'typename seqan::Reference<const seqan::ModifiedString<THost, seqan::Tag<seqan::ModPadding_> > >::Type seqan::value(const seqan::ModifiedString<THost, seqan::Tag<seqan::ModPadding_> >&, TPosition) [with THost = seqan::Segment<seqan::String<seqan::SimpleType<unsigned char, seqan::Dna5_>, seqan::Alloc<> >, seqan::InfixSegment>; TPosition = long unsigned int; typename seqan::Reference<const seqan::ModifiedString<THost, seqan::Tag<seqan::ModPadding_> > >::Type = seqan::SimpleType<unsigned char, seqan::Dna5_>&]':
/home/mi/h4nn3s/devel/lambda/include/seqan/include/seqan/modifier/modifier_string.h:157:21:   required from 'typename seqan::Reference<const seqan::ModifiedString<THost, TSpec> >::Type seqan::ModifiedString<THost, TSpec>::operator[](TPos) const [with TPos = long unsigned int; THost = seqan::Segment<seqan::String<seqan::SimpleType<unsigned char, seqan::Dna5_>, seqan::Alloc<> >, seqan::InfixSegment>; TSpec = seqan::Tag<seqan::ModPadding_>; typename seqan::Reference<const seqan::ModifiedString<THost, TSpec> >::Type = seqan::SimpleType<unsigned char, seqan::Dna5_>&]'
/home/mi/h4nn3s/devel/lambda/include/seqan/include/seqan/align/dp_align_simd_helper.h:102:38:   required from 'void seqan::_createSimdRepImpl(TSimdVecs&, const TStrings&, std::index_sequence<INDICES ...>&) [with TSimdVecs = seqan::String<__vector(8) short int, seqan::Alloc<seqan::Tag<seqan::OverAligned_> > >; TStrings = seqan::StringSet<seqan::ModifiedString<seqan::Segment<seqan::String<seqan::SimpleType<unsigned char, seqan::Dna5_>, seqan::Alloc<> >, seqan::InfixSegment>, seqan::Tag<seqan::ModPadding_> >, seqan::Owner<> >; long unsigned int ...I = {0, 1, 2, 3, 4, 5, 6, 7}; std::index_sequence<INDICES ...> = std::integer_sequence<long unsigned int, 0, 1, 2, 3, 4, 5, 6, 7>]'
/home/mi/h4nn3s/devel/lambda/include/seqan/include/seqan/align/dp_align_simd_helper.h:113:23:   required from 'void seqan::_createSimdRepImpl(TSimdVecs&, const TStrings&) [with TSimdVecs = seqan::String<__vector(8) short int, seqan::Alloc<seqan::Tag<seqan::OverAligned_> > >; TStrings = seqan::StringSet<seqan::ModifiedString<seqan::Segment<seqan::String<seqan::SimpleType<unsigned char, seqan::Dna5_>, seqan::Alloc<> >, seqan::InfixSegment>, seqan::Tag<seqan::ModPadding_> >, seqan::Owner<> >]'
/home/mi/h4nn3s/devel/lambda/include/seqan/include/seqan/align/dp_align_simd_helper.h:243:23:   required from 'void seqan::_prepareAndRunSimdAlignment(TResult&, TTraces&, const TSequencesH&, const TSequencesV&, const TScore&, const seqan::AlignConfig2<TAlgo, TBand, TFreeEndGaps, TTraceback>&, const TGapModel&, seqan::SimdAlignVariableLength<TTraits>) [with TResult = __vector(8) short int; TTraces = seqan::StringSet<seqan::String<seqan::TraceSegment_<unsigned int, long unsigned int>, seqan::Alloc<> >, seqan::Owner<> >; TSequencesH = seqan::Segment<seqan::StringSet<seqan::Segment<seqan::String<seqan::SimpleType<unsigned char, seqan::Dna5_>, seqan::Alloc<> >, seqan::InfixSegment>, seqan::Owner<> >, seqan::InfixSegment>; TSequencesV = seqan::Segment<seqan::StringSet<seqan::Segment<seqan::String<seqan::SimpleType<unsigned char, seqan::Dna5_>, seqan::Alloc<> >, seqan::InfixSegment>, seqan::Owner<> >, seqan::InfixSegment>; TScore = seqan::Score<__vector(8) short int, seqan::ScoreSimdWrapper<seqan::Score<int, seqan::Tag<seqan::Simple_> > > >; TAlgo = seqan::LocalAlignment_<>; TBand = seqan::DPBandConfig<seqan::Tag<seqan::BandOff_> >; TFreeEndGaps = seqan::FreeEndGaps_<seqan::True, seqan::True, seqan::True, seqan::True>; TTraceback = seqan::Tag<seqan::TracebackOff_>; TGapModel = seqan::Tag<seqan::AffineGaps_>; TTraits = seqan::Tag<seqan::Nothing_>]'
/home/mi/h4nn3s/devel/lambda/include/seqan/include/seqan/align/dp_align_simd_helper.h:278:36:   required from 'void seqan::_prepareAndRunSimdAlignment(TResult&, TTraces&, const TSequencesH&, const TSequencesV&, const TScore&, const seqan::AlignConfig2<TAlgo, TBand, TFreeEndGaps, TTraceback>&, const TGapModel&) [with TResult = __vector(8) short int; TTraces = seqan::StringSet<seqan::String<seqan::TraceSegment_<unsigned int, long unsigned int>, seqan::Alloc<> >, seqan::Owner<> >; TSequencesH = seqan::Segment<seqan::StringSet<seqan::Segment<seqan::String<seqan::SimpleType<unsigned char, seqan::Dna5_>, seqan::Alloc<> >, seqan::InfixSegment>, seqan::Owner<> >, seqan::InfixSegment>; TSequencesV = seqan::Segment<seqan::StringSet<seqan::Segment<seqan::String<seqan::SimpleType<unsigned char, seqan::Dna5_>, seqan::Alloc<> >, seqan::InfixSegment>, seqan::Owner<> >, seqan::InfixSegment>; TScore = seqan::Score<__vector(8) short int, seqan::ScoreSimdWrapper<seqan::Score<int, seqan::Tag<seqan::Simple_> > > >; TAlgo = seqan::LocalAlignment_<>; TBand = seqan::DPBandConfig<seqan::Tag<seqan::BandOff_> >; TFreeEndGaps = seqan::FreeEndGaps_<seqan::True, seqan::True, seqan::True, seqan::True>; TTraceback = seqan::Tag<seqan::TracebackOff_>; TGapModel = seqan::Tag<seqan::AffineGaps_>]'
/home/mi/h4nn3s/devel/lambda/src/search_algo.hpp:2311:36:   [ skipping 2 instantiation contexts, use -ftemplate-backtrace-limit=0 to disable ]
/home/mi/h4nn3s/devel/lambda/src/search_algo.hpp:2653:38:   required from 'int iterateMatches(TLocalHolder&) [with TLocalHolder = LocalDataHolder<GlobalDataHolder<seqan::SimpleType<unsigned char, seqan::Dna5_>, seqan::BidirectionalIndex<seqan::FMIndex<void, LambdaFMIndexConfigInBi> >, seqan::Tag<seqan::BlastReport_>, (seqan::BlastProgram)0, (seqan::BlastTabularSpec)0>, seqan::Tag<seqan::AffineGaps_> >]'
/home/mi/h4nn3s/devel/lambda/src/search.hpp:564:37:   required from 'void realMain(LambdaOptions&, const TOutFormat&, const seqan::BlastTabularSpecSelector<h>&, const seqan::BlastProgramSelector<p>&, const TRedAlph&, const TScoreExtension&) [with TIndexSpec = seqan::BidirectionalIndex<seqan::FMIndex<void, LambdaFMIndexConfigInBi> >; TRedAlph = seqan::SimpleType<unsigned char, seqan::Dna5_>; TScoreExtension = seqan::Tag<seqan::AffineGaps_>; TOutFormat = seqan::Tag<seqan::BlastReport_>; seqan::BlastProgram p = (seqan::BlastProgram)0; seqan::BlastTabularSpec h = (seqan::BlastTabularSpec)0]'
/home/mi/h4nn3s/devel/lambda/src/search.hpp:412:60:   required from 'void argConv4(LambdaOptions&, const TOutFormat&, const seqan::BlastTabularSpecSelector<h>&, const seqan::BlastProgramSelector<p>&, const TRedAlph&, const TScoreExtension&) [with TOutFormat = seqan::Tag<seqan::BlastReport_>; TRedAlph = seqan::SimpleType<unsigned char, seqan::Dna5_>; TScoreExtension = seqan::Tag<seqan::AffineGaps_>; seqan::BlastTabularSpec h = (seqan::BlastTabularSpec)0; seqan::BlastProgram p = (seqan::BlastProgram)0]'
/home/mi/h4nn3s/devel/lambda/src/search.hpp:380:24:   required from 'void argConv3(LambdaOptions&, const TOutFormat&, const seqan::BlastTabularSpecSelector<h>&, const seqan::BlastProgramSelector<p>&, const TRedAlph&) [with TOutFormat = seqan::Tag<seqan::BlastReport_>; TRedAlph = seqan::SimpleType<unsigned char, seqan::Dna5_>; seqan::BlastTabularSpec h = (seqan::BlastTabularSpec)0; seqan::BlastProgram p = (seqan::BlastProgram)0]'
/home/mi/h4nn3s/devel/lambda/src/search.hpp:281:28:   required from 'void argConv1(LambdaOptions&, const TOutFormat&, const seqan::BlastTabularSpecSelector<h>&) [with TOutFormat = seqan::Tag<seqan::BlastReport_>; seqan::BlastTabularSpec h = (seqan::BlastTabularSpec)0]'
/home/mi/h4nn3s/devel/lambda/src/search.hpp:259:106:   required from here
/home/mi/h4nn3s/devel/lambda/include/seqan/include/seqan/modifier/modifier_padding.h:230:75: error: binding reference of type 'seqan::Reference<seqan::String<seqan::SimpleType<unsigned char, seqan::Dna5_>, seqan::Alloc<> > >::Type {aka seqan::SimpleType<unsigned char, seqan::Dna5_>&}' to 'const TValue {aka const seqan::SimpleType<unsigned char, seqan::Dna5_>}' discards qualifiers
     return (SEQAN_LIKELY(pos < static_cast<TPosition>(length(host(me))))) ? value(host(me), pos) : cargo(me)._paddedValue;
```
Any ideas, @rrahn ?

--- a/include/seqan/modifier/modifier_padding.h
+++ b/include/seqan/modifier/modifier_padding.h
@@ -165,7 +165,7 @@ struct DefaultIteratorSpec< ModifiedString<THost, ModPadding> >
  * @param [in,out] str  The modified string to be padded.
  * @param [in]     size The number of padded characters.
  * @param [in]     pad  The character to pad the sequence with.
- * 
+ *
  * @datarace Not thread-safe.
  */
 
@@ -211,6 +211,51 @@ length(ModifiedString<THost, ModPadding> const & me)
 }
 
 // ----------------------------------------------------------------------------
+// Function cargoValue()
+// ----------------------------------------------------------------------------
+
+template <typename THost>
+inline typename Reference<ModifiedString<THost, ModPadding> >::Type
+cargoValue(ModifiedString<THost, ModPadding> & me)
+{
+    return cargo(me)._paddedValue;
+}
+
+// NOTE(rrahn): The problem with the padding symbol is, that it is always stored as a member
+// of the modifier class. Hence, if the modifier is const all it's members are const.
+// Now, the cargo could be either defined mutable or, and this what we did right now, the
+// the const is cast-away. However, we use SFINAE to only apply this hack to the Host types,
+// for which this becomes relevant. That are Host types like the Segment class who copy pointer semantics, i.e.
+// the constness of the object is not propagated to the underlying source.
+
+// The default version, where Reference<THost const>::Type gives back a const reference.
+template <typename THost,
+          std::enable_if_t<std::is_same<std::remove_reference_t<
+                                            typename Reference<ModifiedString<THost, ModPadding>>::Type>,
+                                        std::add_const_t<std::remove_reference_t<
+                                            typename Reference<ModifiedString<THost, ModPadding>>::Type>>>::value,
+                           int> = 0>
+inline typename Reference<ModifiedString<THost, ModPadding> >::Type
+cargoValue(ModifiedString<THost, ModPadding> const & me)
+{
+    return cargo(me)._paddedValue;
+}
+
+// The version, where Reference<THost const>::Type gives back a non-const reference.
+template <typename THost,
+          std::enable_if_t<!std::is_same<std::remove_reference_t<
+                                            typename Reference<ModifiedString<THost, ModPadding>>::Type>,
+                                         std::add_const_t<std::remove_reference_t<
+                                            typename Reference<ModifiedString<THost, ModPadding>>::Type>>>::value,
+                           int> = 0>
+inline typename Reference<ModifiedString<THost, ModPadding> >::Type
+cargoValue(ModifiedString<THost, ModPadding> const & me)
+{
+    using TTargetType = typename Reference<ModifiedString<THost, ModPadding> >::Type;
+    return const_cast<TTargetType>(cargo(me)._paddedValue);
+}
+
+// ----------------------------------------------------------------------------
 // Function value()
 // ----------------------------------------------------------------------------
 
@@ -219,7 +264,7 @@ inline typename Reference<ModifiedString<THost, ModPadding> >::Type
 value(ModifiedString<THost, ModPadding> & me, TPosition const pos)
 {
     SEQAN_ASSERT_LT(pos, static_cast<TPosition>(length(me)));
-    return (SEQAN_LIKELY(pos < static_cast<TPosition>(length(host(me))))) ? host(me)[pos] : cargo(me)._paddedValue;
+    return (SEQAN_LIKELY(pos < static_cast<TPosition>(length(host(me))))) ? host(me)[pos] : cargoValue(me);
 }
 
 template <typename THost, typename TPosition>
@@ -227,7 +272,7 @@ inline typename Reference<ModifiedString<THost, ModPadding> const>::Type
 value(ModifiedString<THost, ModPadding> const & me, TPosition const pos)
 {
     SEQAN_ASSERT_LT(pos, static_cast<TPosition>(length(me)));
-    return (SEQAN_LIKELY(pos < static_cast<TPosition>(length(host(me))))) ? value(host(me), pos) : cargo(me)._paddedValue;
+        return (SEQAN_LIKELY(pos < static_cast<TPosition>(length(host(me))))) ? value(host(me), pos) : cargoValue(me);
 }
 
 // ----------------------------------------------------------------------------

--- a/tests/modifier/test_modifier.cpp
+++ b/tests/modifier/test_modifier.cpp
@@ -42,7 +42,7 @@
 #include "test_modifier_string_padding.h"
 
 
-SEQAN_BEGIN_TESTSUITE(test_modifier) 
+SEQAN_BEGIN_TESTSUITE(test_modifier)
 {
     // Test the modifier shortcuts.
     SEQAN_CALL_TEST(test_modifer_shortcuts_dna_string_reverse);
@@ -130,5 +130,6 @@ SEQAN_BEGIN_TESTSUITE(test_modifier)
     SEQAN_CALL_TEST(test_modified_string_padding_end);
     SEQAN_CALL_TEST(test_modified_string_padding_difference);
     SEQAN_CALL_TEST(test_modified_string_padding_iterator);
+    SEQAN_CALL_TEST(test_modified_string_padding_defect_2190);
 }
 SEQAN_END_TESTSUITE

--- a/tests/modifier/test_modifier_string_padding.h
+++ b/tests/modifier/test_modifier_string_padding.h
@@ -181,4 +181,28 @@ SEQAN_DEFINE_TEST(test_modified_string_padding_iterator)
     SEQAN_ASSERT_EQ(*(it - 3), seq[12]);
 }
 
+SEQAN_DEFINE_TEST(test_modified_string_padding_defect_2190)
+{
+    using namespace seqan;
+
+    DnaString seq = "ACGTGGATAGCATCG";
+    auto seqInf = infix(seq, 0, length(seq));
+
+    auto test_const = [](auto const & modifier)
+    {
+        // using TRef = typename Reference<decltype(modifier)>::Type;
+        auto x = value(modifier, 1);
+        SEQAN_ASSERT_EQ(x, 'C');
+    };
+    { // Test working case, when reference of const modifier gives back a const reference to the value
+        ModifiedString<decltype(seq), ModPadding> modString(seq);
+        test_const(modString);
+    }
+
+    {  // Test defect, when reference of const modifier gives back a non-const reference to the value
+        ModifiedString<decltype(seqInf), ModPadding> modString(seqInf);
+        test_const(modString);
+    }
+}
+
 #endif  // #ifndef TESTS_MODIFIER_MODIFIER_STRING_PADDING_H_

